klaus/contrib/auto_klaus.py

```diff
@@ -0,0 +1,103 @@
+"""
+Alternative take on the "automatically discovered repositories" concept
+that requires no threads, polling or inotify. Instead the filesystem is
+consulted whenever a repository name is looked up.
+
+Since os.path.exists() and os.listdir() are fairly quick filesystem
+operations, performance should be good for small to medium sites.
+FancyRepo() objects are cached.
+
+Repositories are identified by the existence of a
+
+    <reponame>/git-daemon-export-ok
+
+file (for compatibility with gitweb).
+
+Example usage:
+
+    from klaus.contrib.auto_klaus import Klaus, SlashDynamicRepos
+
+    application = Klaus('/srv/git', "My git repositories", False)
+
+    application.wsgi_app = httpauth.AlwaysFailingAuthMiddleware(
+        wsgi_app=dulwich.web.make_wsgi_chain(
+            backend=dulwich.server.DictBackend(SlashDynamicRepos(application.valid_repos)),
+            fallback_app=application.wsgi_app,
+        ),
+    )
+"""
+
+import collections.abc
+import pathlib
+import os
+import os.path
+
+import klaus
+import klaus.repo
+
+class FilesystemRepoDict(collections.abc.Mapping):
+    """
+    Maintain a virtual read-only dictionary whose contents represent
+    the presence of git repositories in the given root directory.
+    """
+
+    def __init__(self, root):
+        self._root = pathlib.Path(root)
+        self._repos = {}
+
+    def __getitem__(self, name):
+        if not name or name[0] == '.' or '/' in name or '\0' in name:
+            raise KeyError(name)
+
+        repos = self._repos
+        path = self._root / name
+        if not os.path.exists(path / 'git-daemon-export-ok'):
```

Comment:
I think there's a bug here WRT `.git` suffix handling

If I have a repo named `repo.git` with `repo.git/git-daemon-export-ok`, then it's properly listed on the index page, but the link generated resolves to `/repo` which then 404s because `./repo/git-daemon-export-ok` doesn't exist. `klaus.contrib.autoreload` exhibits the same suffix removal, but correctly resolves the shortened name.

I'm not sure what the best approach to resolve this is, but I think this mapping might need a minor rethingk